octoprint_prusammu/__init__.py

```diff
@@ -343,28 +373,33 @@ def on_event(self, event, payload=None):
     # Fired any time we detect a command that would update something about the MMU
     if event == PluginEventKeys.MMU_CHANGE:
       self._log("on_event {} with".format(event), obj=payload, debug=True)
+      # Fix payload to add missing values
       if MmuKeys.STATE not in payload:
         payload[MmuKeys.STATE] = self.mmu[MmuKeys.STATE]
       if MmuKeys.TOOL not in payload:
         payload[MmuKeys.TOOL] = self.mmu[MmuKeys.TOOL]
       if MmuKeys.PREV_TOOL not in payload:
         payload[MmuKeys.PREV_TOOL] = self.mmu[MmuKeys.PREV_TOOL]
-      if MmuKeys.ERROR not in payload and payload[MmuKeys.STATE] == MmuStates.ATTENTION:
-        payload[MmuKeys.ERROR] = self.mmu[MmuKeys.ERROR]
+      if MmuKeys.RESPONSE not in payload:
+        payload[MmuKeys.RESPONSE] = self.mmu[MmuKeys.RESPONSE]
+      if MmuKeys.RESPONSE_DATA not in payload:
+        payload[MmuKeys.RESPONSE_DATA] = self.mmu[MmuKeys.RESPONSE_DATA]
       # failsafe so we dont constantly spam state changes
       if (
         self.mmu[MmuKeys.STATE] == payload[MmuKeys.STATE] and
         self.mmu[MmuKeys.TOOL] == payload[MmuKeys.TOOL] and
-        self.mmu[MmuKeys.PREV_TOOL] == payload[MmuKeys.PREV_TOOL]
+        self.mmu[MmuKeys.PREV_TOOL] == payload[MmuKeys.PREV_TOOL] and
+        self.mmu[MmuKeys.RESPONSE] == payload[MmuKeys.RESPONSE] and
+        self.mmu[MmuKeys.RESPONSE_DATA] == payload[MmuKeys.RESPONSE_DATA]
       ):
         return
       self.mmu = dict(
-        liveState=self.mmu[MmuKeys.LIVE_STATE],
-        state=None if MmuKeys.STATE not in payload else payload[MmuKeys.STATE],
-        tool=None if MmuKeys.TOOL not in payload else payload[MmuKeys.TOOL],
-        previousTool=None if MmuKeys.PREV_TOOL not in payload else payload[MmuKeys.PREV_TOOL],
-        # Remove error if we are no longer in attention
-        error="" if MmuKeys.ERROR not in payload or payload[MmuKeys.STATE] != MmuStates.ATTENTION else payload[MmuKeys.ERROR])
+        state=payload[MmuKeys.STATE],
+        lastLine=self.mmu[MmuKeys.LAST_LINE],

```

Comment:
Similar to LIVE_STATE, LAST_LINE isn't used for anything down the way, so since we're updating self.mmu here, It's better to pass in the current LAST_LINE value than one that might be in the payload.

As a side note, I now see how LIVE_STATE worked for checks. I currently use several self.mmu[MmuKeys.STATE] checks in gcode_received_hook, and gcode_sent_hook already used self.mmu[MmuKeys.STATE] and self.mmu[MmuKeys.TOOL]. If I'm understanding this right, these checks could fail if the corresponding event to update self.mmu hasn't fired yet, right?

I could add back LIVE_STATE, but that doesn't fix the TOOL check in gcode_sent_hook. How do you feel about updating self.mmu in _fire_event instead of in on_event, so self.mmu is synchronous? We would then pass self.mmu as the payload to on_event, and don't need to update self.mmu there.

I understand that we want to avoid extra processing time as much as possible here. All the MMU3 response checks are deduped, so they only trigger _fire_event once per new MMU response that is different, and there's some other logic in there to reduce it more when not needed. The other two checks I added are single lines that don't spam in the log, so should only fire once. All that remains is the MMU2 portion of the code. PAUSED_USER definitely spams, but with self.mmu being synchronous, the spamminess reduction in gcode_recieved_hook should prevent it from firing _fire_event more than once. The other MMU2 commands would need looked into to see if they spammy.

Reply:
Hm, I was thinking about this today. I think the right direction is to move the logic into _fire_event. There isn't much logic happening so I think we can afford it, and it'll prevent any race conditions. It's a better move than re-introducing LIVE_STATE, which in hindsight i should have done that in the first place. Pre 3.0.0 the commands came in so sparse that it wasn't a problem.

Reply:
I'm also ok merging this as is and before I cut a new release, I can do that refactoring. Happy to let you handle it in this pr, but it didn't feel right asking you to undo my mistake lol.

Reply:
I think I'll let you handle it. Moving the stuff into _fire_event means adding some extra deduplication, and possibly making MMU_CHANGED useless, so I'll let you decide how you want to go about with it. 
Thanks for the quick responses and review!

Reply:
Sounds good! Leme merge it from here, I'll do the refactoring and can do the testing with that version. Appreciate the work! I'll be sure to credit you.